pkg/servicemesh/control-plane.go

```diff
@@ -0,0 +1,578 @@
+
+	return builder
+}
+
+// WithGrafanaAddon adds grafana addons to the serviceMeshControlPlane.
+func (builder *ControlPlaneBuilder) WithGrafanaAddon(
+	enablement bool,
+	install *istioV2.GrafanaInstallConfig,
+	address *string) *ControlPlaneBuilder {
+	if valid, _ := builder.validate(); !valid {
+		return builder
+	}
+
+	glog.V(100).Infof(
+		"Creating serviceMeshControlPlane %s with the Grafana addons defined: enablement %s, install %s, "+
+			"address %s", builder.Definition.Name, enablement, install, address)
+
+	grafanaAddon := &istioV2.GrafanaAddonConfig{
+		Enablement: istioV2.Enablement{
+			Enabled: &enablement,
+		},
+		Install: install,
+		Address: address,
+	}
+
+	if builder.Definition.Spec.Addons == nil {
+		builder.Definition.Spec.Addons = &istioV2.AddonsConfig{}
+	}
+
+	if builder.Definition.Spec.Addons.Grafana == nil {
+		builder.Definition.Spec.Addons.Grafana = &istioV2.GrafanaAddonConfig{}
+	}
+
+	builder.Definition.Spec.Addons.Grafana = grafanaAddon
+
+	return builder
+}
+
+// WithJaegerAddon add joeger addons to the serviceMeshControlPlane.
+func (builder *ControlPlaneBuilder) WithJaegerAddon(
+	name string,
+	elasticsearchRedundancyPolicy string,
+	storageType istioV2.JaegerStorageType,
+	memoryStorageMaxTraces int64,
+	elasticSearchNodesCount int32,
+	elasticsearchStorage istioV1.HelmValues,
+	elasticsearchIndexCleaner istioV1.HelmValues,
+	ingressEnablement bool,
+	ingressLabels map[string]string,
+	ingressAnnotations map[string]string,
+) *ControlPlaneBuilder {
+	if valid, _ := builder.validate(); !valid {
+		return builder
+	}
+
+	glog.V(100).Infof(
+		"Creating serviceMeshControlPlane %s with the Jaeger addons defined", builder.Definition.Name)
+
+	var jaegerAddon = &istioV2.JaegerAddonConfig{
+		Name: name,
+		Install: &istioV2.JaegerInstallConfig{
+			Storage: &istioV2.JaegerStorageConfig{
+				Type: storageType,
+				Memory: &istioV2.JaegerMemoryStorageConfig{
+					MaxTraces: &memoryStorageMaxTraces,
+				},
+				Elasticsearch: &istioV2.JaegerElasticsearchStorageConfig{
+					NodeCount:        &elasticSearchNodesCount,
+					Storage:          &elasticsearchStorage,
+					RedundancyPolicy: elasticsearchRedundancyPolicy,
+					IndexCleaner:     &elasticsearchIndexCleaner,
+				},
+			},
+			Ingress: &istioV2.JaegerIngressConfig{
+				Enablement: istioV2.Enablement{
+					Enabled: &ingressEnablement,
+				},
+				Metadata: &istioV2.MetadataConfig{
+					Labels:      ingressLabels,
+					Annotations: ingressAnnotations,
+				},
+			},
+		},
+	}
+
+	if builder.Definition.Spec.Addons == nil {
+		builder.Definition.Spec.Addons = &istioV2.AddonsConfig{}
+	}
+
+	if builder.Definition.Spec.Addons.Jaeger == nil {
+		builder.Definition.Spec.Addons.Jaeger = &istioV2.JaegerAddonConfig{}
+	}
+
+	builder.Definition.Spec.Addons.Jaeger = jaegerAddon
+
+	return builder
+}
+
+// WithKialiAddon add kiali addons to the serviceMeshControlPlane.
+func (builder *ControlPlaneBuilder) WithKialiAddon(
```

Comment:
Can you pass:
install *istioV2.KialiAddonConfig here like you did in the WithGrafanaAddon function?
It would reduce the number of arguments dramatically and the only verification you'll need to run is:

	if KialiAddonConfig == nil {
		glog.V(100).Infof("The KialiAddonConfig of ServiceMeshControlPlane can not be empty")

		builder.errorMsg = "'KialiAddonConfig' can not be empty for ServiceMeshControlPlane struct"
	}


You can built KialiAddonConfig in your test cases for now and later we will implement kialiaddonconfig.go struct in this package

Reply:
why?

Reply:
please check my comment: https://github.com/openshift-kni/eco-goinfra/pull/147/files#r1362954188